distribution_metadata.c

```diff
@@ -524,62 +467,68 @@ ColumnNameToColumn(Oid relationId, char *columnName)
 
 
 /*
- * LoadShardIntervalRow finds the row for the specified shard identifier in the
- * shard table and copies values from that row into the provided output params.
+ * TupleToShardPlacement populates a ShardInterval using values from a row of
+ * the shard configuration table and returns a pointer to that struct. The
+ * input tuple must not contain any NULLs.
+ *
+ * Because the partition type and column must be known in order to interpret
+ * the min/max values of a shard interval, the provided HeapTuple must contain
+ * these fields. This is easily accomplished by JOINing to the partition table.
  */
-static void
-LoadShardIntervalRow(int64 shardId, Oid *relationId, char **minValue,
-					 char **maxValue)
+static ShardInterval *
+TupleToShardInterval(HeapTuple heapTuple, TupleDesc tupleDescriptor)
 {
-	RangeVar *heapRangeVar = NULL;
-	RangeVar *indexRangeVar = NULL;
-	Relation heapRelation = NULL;
-	Relation indexRelation = NULL;
-	IndexScanDesc indexScanDesc = NULL;
-	const int scanKeyCount = 1;
-	ScanKeyData scanKey[scanKeyCount];
-	HeapTuple heapTuple = NULL;
-
-	heapRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_TABLE_NAME, -1);
-	indexRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_PKEY_INDEX_NAME, -1);
-
-	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
-	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);
+	ShardInterval *shardInterval = NULL;
+	bool isNull = false;
+	Oid intervalTypeId = InvalidOid;
+	int32 intervalTypeMod = -1;
+	Oid inputFunctionId = InvalidOid;
+	Oid typeIoParam = InvalidOid;
 
-	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(shardId));
+	Datum idDatum = SPI_getbinval(heapTuple, tupleDescriptor, 1, &isNull);
```

Comment:
> The number is presumably the field location? This makes me a bit uncomfortable, since we are coupled to the way the query is defined in a different function

Yeah… in places that I specify a query and extract fields in the same function it felt fine, but here I agree it's kind of weird.

> How can we make the two consistent?

I don't quite get this. What's inconsistent?

> I'd rather the query just returns the fields of the table.

I wrote it that way at first and it was quite a bit less clear.

> Maybe we could then use the relation tupleDescriptor (instead of query), so we know what this function expects?

We aren't interacting with relations: we're interacting with queries. Using a relation's `tupleDescriptor` is conceptually wrong. I was careful to _always_ specify column lists in `INSERT` and `SELECT` statements so the logic is entirely unambiguous. Using a relation's descriptor with e.g. `SELECT *` relies on some assumptions I'd rather not make.

> Is there a downside to sticking with the PartitionColumn function to load the column and its type like we did before? (Instead of joining with partition).

Yeah: recursive use of SPI requires some boilerplate calls and involves more reasoning about memory contexts, etc. I avoided any recursive SPI calls in this pull request to keep away from all that.

Also, the code is less straightforward and self-similar that way. As of this pull request, the shard interval and shard placement paths are structurally identical: they run a query and map rows to objects. The old shard interval path was weirder.